### PR TITLE
gh-119727: Add --single-process option to regrtest

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-119727

### CODE REVIEW

**Comment by vstinner** [comment]
cc @gpshead @Yhg1s

File: Lib/test/libregrtest/cmdline.py [hunk 1]
```diff
@@ -307,6 +308,10 @@ def _create_parser():
     group.add_argument('-j', '--multiprocess', metavar='PROCESSES',
                        dest='use_mp', type=int,
                        help='run PROCESSES processes at once')
+    group.add_argument('--sequentially', action='store_true',
```
Comment (serhiy-storchaka) [inline Lib/test/libregrtest/cmdline.py]: You can also implement it as `action='store_const', dest='use_mp', const=None`.

File: Lib/test/libregrtest/worker.py [hunk 2]
```diff
@@ -14,9 +14,9 @@
 
 
 USE_PROCESS_GROUP = (hasattr(os, "setsid") and hasattr(os, "killpg"))
-NEED_TTY = set('''
-    test_ioctl
-'''.split())
+NEED_TTY = {
```
Comment (serhiy-storchaka) [inline Lib/test/libregrtest/worker.py:17]: Not related.

**Review by serhiy-storchaka — APPROVED** [review]
Alternative names are `--singleprocess` and `--no-multiprocess`, which more accurately describe the effect, but this is up to you.

I think that the implementation can be simplified by reusing the same destination variable. It is more common in such cases.

**Comment by vstinner** [comment]
@Yhg1s @hugovk: Since the release process requires to run tests sequentially, do you have a preference for the option name?

* `--singleprocess` (close to the existing `--single` option)
* `--no-multiprocess`
* `--sequentially`
* `--sameprocess`
* something else?

**Comment by hugovk** [comment]
My first thought is to use something like `-j 1` rather than adding a new option. Would this work?

```
  -j, --multiprocess PROCESSES
                        run PROCESSES processes at once
```

This is similar to pytest-xdist (`-n` or `--numprocesses`) and sphinx-build (`-j` or `--jobs`).

**Comment by vstinner** [comment]
> My first thought is to use something like -j 1 rather than adding a new option. Would this work?

Currently, -j1 runs tests sequentially but spawn a new process for each test file. I'm open to consider changing this behavior.

**Comment by hugovk** [comment]
Do you think anyone will realistically want to run sequentially and spawn a new process for each? My guess is not.

**Comment by vstinner** [comment]
I'm a little bit worried that some projects currently run tests with `-j1` with good reasons, and changing that can lead to surprising results. I would prefer to add a new option.

According to the discussion, the least bad choice is `--singleprocess` option name :-)

**Comment by vstinner** [comment]
I updated the PR to use `--single-process` name. Are you ok with that @serhiy-storchaka and @hugovk?

**Comment by serhiy-storchaka** [comment]
It does not really matter. It was just a tiny nitpick from my side. Use the name that you like if you are sure that it will not create problems in future.

File: Lib/test/libregrtest/worker.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/libregrtest/worker.py:17]: Yeah, it's a follow-up of a previous PR. I didn't want to write a dedicated PR for that.

File: Misc/NEWS.d/next/Tests/2024-05-29-15-28-08.gh-issue-119727.dVkaZM.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Add ``--sequentially`` command line option to Python test runner (regrtest).
```
Comment (hugovk) [inline Misc/NEWS.d/next/Tests/2024-05-29-15-28-08.gh-issue-119727.dVkaZM.rst]:
```suggestion
Add ``--single-process`` command line option to Python test runner (regrtest).
```

And update the PR title.

**Review by hugovk — APPROVED** [review]

**Comment by mhsmith** [comment]
I'll submit a backport of this because it would be useful for the iOS and Android buildbots – https://github.com/python/cpython/pull/122992#issuecomment-2287554733.

**Pull request merged**